### PR TITLE
BZ-2010564 Creating initial submit - configuring seccomp profile

### PR DESCRIPTION
**Pull request by kquinn1204**
BZ-2010564: Lack of description about creating and configuring custom SCC that allows configuring seccomp profile

Version(s):
  * PR applies to multiple specific versions:  4.11 and greater 

Issue:
https://bugzilla.redhat.com/show_bug.cgi?id=2010564

Link to docs preview:
http://file.emea.redhat.com/kquinn/BZ-2010564/security/seccomp-profiles.html

### CODE REVIEW

**Comment by kquinn1204** [comment]
HI @kashyap-nirupma can you take a look at this PR which is trying to address issues raised in https://issues.redhat.com/browse/OSDOCS-3736

File: modules/configuring-default-seccomp-profile.adoc [hunk 1]
```diff
@@ -1,15 +1,99 @@
...
+          ],
+          "default": true,
+          "dns": {}
+      }]
+    openshift.io/deployment-config.latest-version: "1"
+    openshift.io/deployment-config.name: parksmap
+    openshift.io/deployment.name: parksmap-1
+    openshift.io/generated-by: OpenShiftWebConsole
+    openshift.io/scc: restricted-v2 <1>
+    seccomp.security.alpha.kubernetes.io/pod: runtime/default <2>
```
Comment (mvazquezc) [inline modules/configuring-default-seccomp-profile.adoc:71]: Instead of showing this, it may be better to show the pod's/containers securityContext information where the seccomp is configured. These annotations are deprecated iirc.

File: modules/configuring-default-seccomp-profile.adoc [hunk 2]
```diff
@@ -1,15 +1,99 @@
...
+---- 
+<1> The `restricted-v2` is set by default on newly created pods in {product-version}. 
+<2> The `restricted-v2` sets the seccompProfile to `runtime/default` by default. 
+
+You can enable the default seccomp profile for a pod or container workload that does not have the seccompProfile set by adding `restricted-v2` to the default service account.
```
Comment (mvazquezc) [inline modules/configuring-default-seccomp-profile.adoc:76]: This is incorrect. All authenticated users (this includes ServiceAccounts) have access to the restricted-v2 SCC.

File: modules/configuring-default-seccomp-profile.adoc [hunk 3]
```diff
@@ -1,15 +1,99 @@
...
+    seccomp.security.alpha.kubernetes.io/pod: runtime/default <2>
+---- 
+<1> The `restricted-v2` is set by default on newly created pods in {product-version}. 
+<2> The `restricted-v2` sets the seccompProfile to `runtime/default` by default. 
+
+You can enable the default seccomp profile for a pod or container workload that does not have the seccompProfile set by adding `restricted-v2` to the default service account. 
+
+Follow these steps:
+
+. Add the `restricted-v2` to the ServiceAccount:
```
Comment (mvazquezc) [inline modules/configuring-default-seccomp-profile.adoc:80]: As mentioned above, this is not required, all authenticated users have access to the SCC.

File: modules/configuring-default-seccomp-profile.adoc [hunk 4]
```diff
@@ -1,15 +1,99 @@
...
+. Add the `restricted-v2` to the ServiceAccount: 
++
+[source, terminal]
+----
+$ oc adm policy add-scc-to-user restricted-v2 -z default
+----
++
+[NOTE]
+====
+The `default` service account is the one taking the responsibility of running the pods, and OpenShift uses and injects this service account into every pod that is launched.
```
Comment (mvazquezc) [inline modules/configuring-default-seccomp-profile.adoc:89]: This is incorrect. That ServiceAccount will be used unless the user configures a different one. Also, it's not OpenShift injecting it. This is default K8s behavior, nothing specific for OCP.

File: modules/configuring-default-seccomp-profile.adoc [hunk 5]
```diff
@@ -1,15 +1,99 @@
...
+.Expected output
++
+[source, terminal]
 ----
-spec:
-  securityContext:
-    seccompProfile:
-      type: RuntimeDefault
+clusterrole.rbac.authorization.k8s.io/system:openshift:scc:restricted-v2 added: "default"
 ----
```
Comment (mvazquezc) [inline modules/configuring-default-seccomp-profile.adoc:98]: I will keep an example showing how to configure a different seccomp profile. You can get ideas from here https://github.com/mvazquezc/scc-fun/blob/main/demo2/README.md#using-the-security-profiles-operator

File: modules/configuring-default-seccomp-profile.adoc (hunk 4 above)
Comment (mvazquezc) [reply modules/configuring-default-seccomp-profile.adoc:89]: What OCP will do though is configure the seccomp profile of the pod based on the information in the SCC (restricted-v2 in this case).

File: security/seccomp-profiles.adoc [hunk 6]
```diff
@@ -7,13 +7,11 @@ include::_attributes/common-attributes.adoc[]
...
 An {product-title} container or a pod runs a single application that performs one or more well-defined tasks. The application usually requires only a small subset of the underlying operating system kernel APIs.
-Seccomp, secure computing mode, is a  Linux kernel feature that can be used to limit the process running in a container to only call a subset of the available system calls. These system calls can be configured by creating a profile that is applied to a container or pod.
-Seccomp profiles are stored as JSON files on the disk.
+Seccomp, secure computing mode, is a  Linux kernel feature that can be used to limit the process running in a container to only call a subset of the available system calls. 
 
-[IMPORTANT]
-====
-OpenShift workloads run unconfined by default, without any seccomp profile applied.
-====
+The `restricted-v2` SCC applies to all newly created pods in {product-version}. The default seccomp profile `runtime/default` is applied to these pods.
```
Comment (mvazquezc) [inline security/seccomp-profiles.adoc:12]: This is not strictly correct. I would reword it to something like:

The `restricted-v2` SCC is the new SCC by default, if your workload doesn't have access to a different SCC, newly created pods in {product-version} will have its seccomp profile configured to `runtime/default` as mandated by the SCC.

**Review by mvazquezc — COMMENTED** [review]
Added some comments.

File: modules/configuring-default-seccomp-profile.adoc (hunk 2 above)
Comment (kquinn1204) [reply modules/configuring-default-seccomp-profile.adoc:76]: So @mvazquezc if I have a pod that was running on 4.10 and I upgraded to 4.11 it will have seccompProfile set to `runtime/default` by default is that correct. I tried to do a few tests. I maybe missing something subtle here and maybe I am showing my lack of knowledge but "All authenticated users (this includes ServiceAccounts) have access to the restricted-v2 SCC" does that mean the same as all running pods will have seccompProfile set to `runtime/default` by default. Here I am trying to address the situation where a pod may not have it set but maybe this is not a valid case now in 4.11.

File: modules/configuring-default-seccomp-profile.adoc (hunk 1 above)
Comment (kquinn1204) [reply modules/configuring-default-seccomp-profile.adoc:71]: @mvazquezc Here I am simply trying to illustrate that this new restricted-v2 is applied. I added the apps per this guidance https://redhat-scholars.github.io/openshift-starter-guides/rhs-openshift-starter-guides/parksmap-container-image.html#deploy_your_first_image like a test pod really. So maybe these annotations are not deprecated this is the output I captured.

File: modules/configuring-default-seccomp-profile.adoc (hunk 3 above)
Comment (kquinn1204) [reply modules/configuring-default-seccomp-profile.adoc:80]: Hi @mvazquezc I guess I need some clarification on "all authenticated users have access to the SCC" what does that mean exactly. All pods have `restricted-v2` set by default

File: security/seccomp-profiles.adoc (hunk 6 above)
Comment (kquinn1204) [reply security/seccomp-profiles.adoc:12]: what is the situation with existing running pods or pods with access to a different SCC for example `restricted`. For these pods I assume if you want to have seccomp profile configured to runtime/default you need to update and that is what I am trying to show how to do here by adding the `restricted-v2` to the ServiceAccount. I am trying to work out how to set the seccomp profile to runtime/default for pods where it is not set already.

File: modules/configuring-default-seccomp-profile.adoc (hunk 1 above)
Comment (mvazquezc) [reply modules/configuring-default-seccomp-profile.adoc:71]: The SCC annotation is valid indeed. I was referring to the `seccomp.security.alpha.kubernetes.io/pod` annotation. I'm okay leaving this at it is.

File: modules/configuring-default-seccomp-profile.adoc (hunk 2 above)
Comment (mvazquezc) [reply modules/configuring-default-seccomp-profile.adoc:76]: Well, there is a subtle difference indeed. 

If a 4.11 cluster was upgraded from 4.10, the default SCC will continue to be `restricted`. 

If it's a 4.11 fresh install, the default SCC will be `restricted-v2`, and pods will get the default seccomp profile configured.

File: modules/configuring-default-seccomp-profile.adoc (hunk 3 above)
Comment (mvazquezc) [reply modules/configuring-default-seccomp-profile.adoc:80]: There is a group in K8s named `system:authenticated`. In OCP 4.10 and below the `restricted` SCC was available for this group. In a fresh install of OCP 4.11, this group will have access to the `restricted-v2` SCC instead.

And what SCC a pod gets assigned depends on the SCCs that the user creating that pod has access to, you can read more about that [here](https://docs.google.com/presentation/d/1mwqYQP-HRaCgk317F5IIlvhRYfVBC5lKbFSu5KATXmM/edit#slide=id.gbca891d708_0_11).

File: security/seccomp-profiles.adoc (hunk 6 above)
Comment (mvazquezc) [reply security/seccomp-profiles.adoc:12]: As mentioned earlier we may need to have two sections now.

4.11 clusters that were upgraded from 4.10
4.11 fresh installed clusters

For the former, you're right. For the later we should reword as I suggested.

File: modules/configuring-default-seccomp-profile.adoc [hunk 7]
```diff
@@ -70,13 +70,14 @@ metadata:
...
 
-You can enable the default seccomp profile for a pod or container workload that does not have the seccompProfile set by adding `restricted-v2` to the default service account. 
-
-Follow these steps:
+The default SCC for: 
 
+* A 4.11 fresh install is `restricted-v2`, and pods get the default seccomp profile `runtime/default` set.
+* A 4.11 cluster upgraded from {product-title} version 4.10 or below is `restricted`. In this case if you want to have the seccomp profile `runtime/default` set on all pods follow the step below: 
++
 . Add the `restricted-v2` to the ServiceAccount:
```
Comment (mvazquezc) [inline modules/configuring-default-seccomp-profile.adoc:81]: This is a tricky one, if the pod can be run with both SCCs, since the last ordering is alphabetical order, the workload will still get assigned to the `restricted` SCC. In that case user will need to force the use of the seccomp profile in the securityContext of the pod or container, which will force the pod to only be admitted by `restricted-v2` (since restricted doesn't allow to configure a seccomp profile).

File: modules/configuring-default-seccomp-profile.adoc [hunk 8]
```diff
@@ -73,6 +73,25 @@ metadata:
 <1> The `restricted-v2` SCC is the SCC added by default, if your workload does not have access to a different SCC. 
 <2> Newly created pods in {product-version} will have the seccomp profile configured to `runtime/default` as mandated by the SCC. 
 
+In clusters upgraded to {product-version} all authenticated users have access to the `restricted` and `restricted-v2` SCC. 
+
+In this scenario if a workload is running on a 4.10 cluster that was admitted by the SCC `restricted`. When upgraded that workload may get admitted by `restricted-v2` if the workload is able to run with `restricted-v2`. That is because `restricted-v2` is the most restrictive SCC between `restricted` and `restricted-v2`.
+Conversely in a cluster upgraded to {product-version} with a workload that requires `privilegeEscalation: true` this workload will continue to have the `restricted` SCC available for any authenticated user.
```
Comment (mvazquezc) [inline modules/configuring-default-seccomp-profile.adoc:79]: I will put a disclaimer saying something like:

In the example below we're using privilegeEscalation as a feature that is allowed by restricted but not by restricted-v2. There are more features that were allowed by restricted and now are denied by restricted-v2.

Comment (kquinn1204) [reply modules/configuring-default-seccomp-profile.adoc]: Hi @mvazquezc I hope what I now updated more accurately reflects the situation. Hopefully close to a LGTM so I can get this content merged.

Comment (mvazquezc) [reply modules/configuring-default-seccomp-profile.adoc]: lgtm!

File: modules/configuring-default-seccomp-profile.adoc [hunk 9]
```diff
@@ -1,15 +1,106 @@
...
+          "dns": {}
+      }]
+    openshift.io/deployment-config.latest-version: "1"
+    openshift.io/deployment-config.name: parksmap
+    openshift.io/deployment.name: parksmap-1
+    openshift.io/generated-by: OpenShiftWebConsole
+    openshift.io/scc: restricted-v2 <1>
+    seccomp.security.alpha.kubernetes.io/pod: runtime/default <2>
+---- 
+<1> The `restricted-v2` SCC is the SCC added by default, if your workload does not have access to a different SCC.
```
Comment (nalhadef) [inline modules/configuring-default-seccomp-profile.adoc:81]: Can this sentence be rewritten so "SCC" isn't used three times?

File: modules/configuring-default-seccomp-profile.adoc [hunk 10]
```diff
@@ -1,15 +1,106 @@
...
+---- 
+<1> The `restricted-v2` SCC is the SCC added by default, if your workload does not have access to a different SCC. 
+<2> Newly created pods in {product-version} will have the seccomp profile configured to `runtime/default` as mandated by the SCC. 
+
+[id="upgraded_cluster_{context}"]
+== Upgraded cluster
+
+In clusters upgraded to {product-version} all authenticated users have access to the `restricted` and `restricted-v2` SCC. 
+
+If a workload is running for example on a {product-title} v4.10 cluster that was admitted by the SCC `restricted` when the cluster is upgraded that workload may get admitted by `restricted-v2` if the workload is able to run with `restricted-v2`. This is because `restricted-v2` is the most restrictive SCC between `restricted` and `restricted-v2`.
```
Comment (nalhadef) [inline modules/configuring-default-seccomp-profile.adoc:89]: The first sentence has 43 words, which is waaaaay to many. Thirty-ish is pushing it. Please rewrite this, breaking it up into two or three sentences, if possible.

Re: second sentence, if there are only two options, you can't use "most" or "least." In this case, you'd want "more restrictive."

https://www.onestopenglish.com/support-for-teaching-grammar/comparative-and-superlative-adjectives-article/144846.article

File: modules/configuring-default-seccomp-profile.adoc [hunk 11]
```diff
@@ -1,15 +1,106 @@
...
+Conversely with a workload that requires `privilegeEscalation: true` this workload will continue to have the `restricted` SCC available for any authenticated user. This is because `restricted-v2` does not allow `privilegeEscalation`.
+
+[id="newly_installed_{context}"]
+== Newly installed cluster
+
+For newly installed {product-title} v4.11 cluster, the `restricted-v2` replaces the `restricted` SCC as an SCC that is available to be used by any authenticated user. A workload with `privilegeEscalation: true`, will not be admitted into the cluster since `restricted-v2` is the only SCC available for authenticated users by default.
```
Comment (nalhadef) [inline modules/configuring-default-seccomp-profile.adoc:95]: Last sentence: Change "...will not be admitted..." to "...is not admitted..."

File: modules/configuring-default-seccomp-profile.adoc [hunk 12]
```diff
@@ -1,15 +1,106 @@
...
+
+If a workload is running for example on a {product-title} v4.10 cluster that was admitted by the SCC `restricted` when the cluster is upgraded that workload may get admitted by `restricted-v2` if the workload is able to run with `restricted-v2`. This is because `restricted-v2` is the most restrictive SCC between `restricted` and `restricted-v2`.
+Conversely with a workload that requires `privilegeEscalation: true` this workload will continue to have the `restricted` SCC available for any authenticated user. This is because `restricted-v2` does not allow `privilegeEscalation`.
+
+[id="newly_installed_{context}"]
+== Newly installed cluster
+
+For newly installed {product-title} v4.11 cluster, the `restricted-v2` replaces the `restricted` SCC as an SCC that is available to be used by any authenticated user. A workload with `privilegeEscalation: true`, will not be admitted into the cluster since `restricted-v2` is the only SCC available for authenticated users by default. 
+
+`privilegeEscalation` is a feature that is allowed by `restricted` but not by `restricted-v2`. It is important to note that more features are denied by `restricted-v2` than were allowed by `restricted` SCC.
```
Comment (nalhadef) [inline modules/configuring-default-seccomp-profile.adoc:97]: - Avoid starting a sentence with lowercase words. Try: "The feature `privilegeEscalation` is ..."
- Second sentence: Delete "It is important to note that," starting the sentence with "More."

File: modules/configuring-default-seccomp-profile.adoc [hunk 13]
```diff
@@ -1,15 +1,106 @@
...
+Conversely with a workload that requires `privilegeEscalation: true` this workload will continue to have the `restricted` SCC available for any authenticated user. This is because `restricted-v2` does not allow `privilegeEscalation`.
+
+[id="newly_installed_{context}"]
+== Newly installed cluster
+
+For newly installed {product-title} v4.11 cluster, the `restricted-v2` replaces the `restricted` SCC as an SCC that is available to be used by any authenticated user. A workload with `privilegeEscalation: true`, will not be admitted into the cluster since `restricted-v2` is the only SCC available for authenticated users by default. 
+
+`privilegeEscalation` is a feature that is allowed by `restricted` but not by `restricted-v2`. It is important to note that more features are denied by `restricted-v2` than were allowed by `restricted` SCC.  
+
+To admit a workload with `privilegeEscalation: true` into a newly installed {product-title} v4.11 cluster you need to give access to the `restricted` SCC to the ServiceAccount running the workload (or any other SCC that can admit this workload) using a RoleBinding. Run the following command to do this:
```
Comment (nalhadef) [inline modules/configuring-default-seccomp-profile.adoc:99]: The first sentence has 43 words. Please break it into small chunks.

File: security/seccomp-profiles.adoc [hunk 14]
```diff
@@ -7,13 +7,11 @@ include::_attributes/common-attributes.adoc[]
 toc::[]
 
 An {product-title} container or a pod runs a single application that performs one or more well-defined tasks. The application usually requires only a small subset of the underlying operating system kernel APIs.
-Seccomp, secure computing mode, is a  Linux kernel feature that can be used to limit the process running in a container to only call a subset of the available system calls. These system calls can be configured by creating a profile that is applied to a container or pod.
-Seccomp profiles are stored as JSON files on the disk.
+Seccomp, secure computing mode, is a  Linux kernel feature that can be used to limit the process running in a container to only call a subset of the available system calls.
```
Comment (nalhadef) [inline security/seccomp-profiles.adoc:10]: - I think this sentence would read better as "Secure computing mode, secomp,..." but it's minor.
- To avoid having "call(s)" used twice, could we end with: "... to only using a subset of the available system calls."

File: modules/configuring-default-seccomp-profile.adoc (hunk 9 above)
Comment (nalhadef) [reply modules/configuring-default-seccomp-profile.adoc]: @kquinn1204 Remove comma after default.
